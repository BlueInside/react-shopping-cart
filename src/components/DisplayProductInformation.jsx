import { useState } from 'react';
import Button from './Button';
import useDataFetching from '../hooks/useDataFetching';
import { useParams } from 'react-router-dom';
import { addToCard } from '../utils/cart';
function DisplayProductInformation() {
  const { productId } = useParams();
  const [quantity, setQuantity] = useState(1);
  const { data, loading, error } = useDataFetching(productId);

  function handleOnQuantityChange(e) {
    const numericValue = Number(e.target.value.replace(/[^0-9]/g, ''));
    if (e.target.value === '') {
      setQuantity(1);
      return;
    }

    if (!isNaN(numericValue)) {
      setQuantity(numericValue);
    } else {
      setQuantity(1);
    }
  }

  function handleIncrement() {
    setQuantity((q) => q + 1);
  }

  function handleReduce() {
    if (quantity - 1 >= 1) {
      // Min quantity is 1
      setQuantity((q) => q - 1);
    } else return;
  }
  if (loading || !data) return <div>Loading</div>;
  if (error) {
    return <div>Oops! Something went wrong. Please try again later.</div>;
  }
  const { title, image, rating, price, category, description } = data;
  console.log(data);

  return (
    <div role="productContainer">
      <div>
        <img src={image} alt="product" role="productImage" />
        <div>
          <div>
            <p role="title">{title}</p>
            {rating && <p role="rating">{rating.rate}</p>}
            <p role="price">{price}</p>
          </div>

          <div>
            <div className="quantityControllers">
              <Button handleClick={handleIncrement} label={'+'} role="add" />
              <input
                role="quantity"
                type="number"
                min={1}
                value={quantity}
                onChange={handleOnQuantityChange}
              />
              <Button handleClick={handleReduce} label={'-'} role={'reduce'} />
            </div>
            <Button
              handleClick={() => addToCard({ ...data }, quantity)}
              role={'addToCart'}
              label={'Add to cart'}
            />
          </div>
        </div>
      </div>
      <div>
        <div>
          <p role="category">{category}</p>
          <p role="productDescription">{description}</p>
        </div>
      </div>
    </div>
  );
}

DisplayProductInformation.defaultProps = {};

export default DisplayProductInformation;
