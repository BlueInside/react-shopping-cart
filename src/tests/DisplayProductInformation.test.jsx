import { beforeEach, describe, expect, it } from 'vitest';
import DisplayProductInformation from '../components/DisplayProductInformation';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

describe('DisplayProductInformation component', () => {
  let product;
  beforeEach(() => {
    product = {
      id: 4,
      title: "Women's Classic Trench Coat",
      price: 89.99,
      description:
        'Stay stylish and comfortable with this classic trench coat for women. Perfect for rainy days and a fashionable choice for any season.',
      category: "women's clothing",
      image: 'https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg',
      rating: { rate: 4.2, count: 150 },
    };
  });

  it('renders DisplayProductInformation component correctly', () => {
    const { container } = render(
      <DisplayProductInformation {...product} handleAddToCart={() => {}} />
    );
    expect(container).toMatchSnapshot();
  });

  it('displays product details accordingly to product object', async () => {
    render(
      <DisplayProductInformation {...product} handleAddToCart={() => {}} />
    );

    expect(screen.getByRole('productImage')).toHaveAttribute(
      'src',
      product.image
    );
    expect(screen.getByRole('productDescription')).toHaveTextContent(
      product.description
    );
    expect(screen.getByRole('title')).toHaveTextContent(product.title);
    expect(screen.getByRole('price')).toHaveTextContent(product.price);
    expect(screen.getByRole('rating')).toHaveTextContent(product.rating.rate);
    expect(screen.getByRole('category')).toHaveTextContent(product.category);
    expect(screen.getByRole('addToCart')).toBeInTheDocument();
  });

  it('let user adjust quantity', async () => {
    const user = userEvent.setup();

    render(
      <DisplayProductInformation {...product} handleAddToCart={() => {}} />
    );
    const quantityInput = screen.getByRole('quantity');
    const addBtn = screen.getByRole('add');
    const reduceBtn = screen.getByRole('reduce');
    // Initial quantity is 1
    expect(quantityInput).toHaveValue(1);

    // Clicking remove when the quantity is already 0 should not decrease it
    await user.click(reduceBtn);
    expect(quantityInput).toHaveValue(1);

    // // Clicking add should increase the quantity to 1
    await user.click(addBtn);
    expect(quantityInput).toHaveValue(2);

    // Clicking add twice should increase the quantity to 3
    await user.click(addBtn);
    await user.click(addBtn);
    expect(quantityInput).toHaveValue(4);

    // When user clears input it should show 1
    await user.clear(quantityInput);
    expect(quantityInput).toHaveValue(1);

    // Checks if inputs updates value correctly
    await user.type(quantityInput, '234');
    expect(quantityInput).toHaveValue(1234);

    // Checks if input deletes numbers properly
    await user.type(quantityInput, '{backspace}{backspace}{backspace}');
    expect(quantityInput).toHaveValue(1);

    // Ignores letters
    await user.type(quantityInput, '234abc');
    expect(quantityInput).toHaveValue(1234);
  });
});
