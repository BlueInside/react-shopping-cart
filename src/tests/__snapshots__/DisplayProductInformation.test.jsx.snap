// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DisplayProductInformation component > renders DisplayProductInformation component correctly 1`] = `
<div>
  <div
    role="productModal"
  >
    <div>
      <img
        alt="product"
        role="productImage"
        src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
      />
      <div>
        <div>
          <p
            role="title"
          >
            Women's Classic Trench Coat
          </p>
          <p
            role="rating"
          >
            4.2
          </p>
          <p
            role="price"
          >
            89.99
          </p>
        </div>
        <div
          class="quantityControllers"
        >
          <button
            role="add"
          >
            +
          </button>
          <span
            role="quantity"
          >
            0
          </span>
          <button
            role="reduce"
          >
            -
          </button>
        </div>
      </div>
    </div>
    <div>
      <div>
        <p
          role="category"
        >
          women's clothing
        </p>
        <p
          role="productDescription"
        >
          Stay stylish and comfortable with this classic trench coat for women. Perfect for rainy days and a fashionable choice for any season.
        </p>
      </div>
    </div>
  </div>
</div>
`;
